Feature/state

emtuls opened the pull request.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.stream.*;
+
+
+import javax.tools.DocumentationTool.Location;
```
Unused import

Sweethezus replied:
I have VS Code for doing this to me

greenca6 replied:
Yea - it can be a pain for auto-imports at times I've noticed. For me it's also slick enough to let me know when I have unused stuff too though which is nice.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+public class Game {
+    private HashSet<Card> winningCards;
+    private Board board;
+    private Turn turnInstance;
```
I think we should rename this to `currentPlayersTurn` so it's more descriptive

Sweethezus replied:
(y)

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+        //TODO: Set the initialPlayers private List to the players
+        //TODO: Set the game state to GameState.IN_PROGRESS
+
+        this.winningCards = new HashSet<>(this.board.drawWinningCards());
```
I think `drawWinningCards()` already returns a `Set`? Which means you don't have to pass this into a `HashSet` instance.

Sweethezus replied:
It does, but it yelled at me without that explicit conversion. Though it could have been something else.

greenca6 replied:
That's interesting. Okay, maybe it's fine as is then.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+        
+        //TODO: Set the current players turn to the player who will go first, as well as other relevant turn
+        //information
+    }
```
So for this constructor, I'd suggest splitting up the initialization logic into a few different `private` methods, which you can then call in here. That should make the logic more legible and less verbose.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+
+        if(!validPlayers.isEmpty()){
+
+            for(Player p: validPlayers){
```
I don't think we want a for-loop here? If we have a valid player that can/needs to make a rebuttal - then that's the _only_ player that should make a rebuttal. I don't think we need to ask for rebuttals for each of the players that have cards that were in the suggestion.

Sweethezus replied:
What's your rationale behind this? Is it because as soon as we get to the first rejection, we stop the suggestion turn?

greenca6 replied:
So the game rules dictate this I think. The rules states something like: "When a player makes a suggestion, the first player to his/her left shall show a card contained in the suggestion to disprove that suggestion, if such a player is found."

So in this code, someone is making the initial suggestion, so we find the first player (if possible) who should make the rebuttal. The `Suggestion` has yet to happen.

Sweethezus replied:
Follow up: is the stream strictly ordered like the list? (i.e. everytime we run with the same parameters, we'll get the same result from the `getFirst` call?

greenca6 replied:
I _believe_ so. It would be very odd behavior if not.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+        
+
+        List<Player> validPlayers = this.playerList.stream().
+        filter(p -> p.hasCard(suggestedWeapon) || p.hasCard(suggestedRoom) || p.hasCard(suggestedPiece)).collect(Collectors.toList());
```
Going off of my other comment regarding the for-loop, I think we can use the `findFirst()` stream method? This would also allow you to use the `isPresent()` method in your if statement below. So the change would look something like:

```java
Optional<Player> playerToRebut = this.playerList.stream().filter(/* query */).findFirst();

if (playerToRebut.isPresent()) {
    // do something with playerToRebut.get()
}
```

Sweethezus replied:
Certainly. If we don't have to loop, Optional makes the most sense.

greenca6 replied:
Oh, other rules here:

1. We want to make sure that when we iterate at the correct starting index. Remember that the players sit in a circle. In this code, we're _always_ starting at the 0th index, but we need to start at the index that the current player is at, and go "around" the circle.
2. We also want to skip the current player in the filter. It could be feasible that the current player is making a suggestion, with a card that he/she holds. But, obviously you can't rebut your own suggestion.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+
+        Card arr[] = {accusedWeapon, accusedLocation, accusedPiece};
+
+        Set<Card> accusedCards = new HashSet<>(Arrays.asList(arr));
```
This is a nifty way of creating an array and turning it into a `Set`. Can it be simplified to:

```java
Set<Card> accusedCards = new HashSet<>(Arrays.asList({accusedWeapon, accusedLocation, accusedPiece}));
```

Not sure if it can be but it could be worth the shot.

greenca6 replied:
Found an even simpler way to do it: https://stackoverflow.com/a/9470339/9257823

Sweethezus replied:
Your google-fu is better than mine.

greenca6 replied:
🐱‍👤

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+        return this;
+    }
+
+    public Game endTurn(Player p) {
```
I think we should add another `private` method, `setCurrentPlayerTurn(Player p)` for two reasons:

1. At the start of the game, we'll need to do all the stuff required to set the current players turn to the first player
2. This method, `endTurn`, will also call that private method

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -0,0 +1,144 @@
+        return this;
+    }
+
+    private Player getNextActivePlayer(Player currentPlayer) {
```
I like it. Although the parameter `currentPlayer` here shouldn't be needed - we should already know who the current player is by the `currentPlayerTurn` instance that this class is holding.

This method should also take into account if this is the _first_ move of the game - in which case that would mean it's Mrs. Scarlett's turn, if someone is controlling her piece.

Sweethezus replied:
I'm not following how the first move of the game would affect this logic in any way, shape or form.

greenca6 replied:
Well if it's the first move, then that means the turn instance we hold, `currentPlayersTurn` would be `null` probably. That means this method will do that null check, then say, "okay, I'll find who the _first_ player to go is". Which has specific rules - if someone is playing with Mrs. Scarlett, that player goes first. Otherwise, we can just let the first person in the `playerList` go first.

Sweethezus replied:
I think we should set up playerList so that they are in correct order (i.e. whomever is supposed to go first will always go first) rather than bake that logic into this function. Further, I don't think currentPlayersTurn should ever be null (because there will always be a turn as long as the game is alive). If anything, maybe we put in a function to get the absolute first player (i.e. array index 0), which we can use at start up and when the list wraps.

greenca6 replied:
When you first set up the `playerList` though, wouldn't you want to know who the first player is? Couldn't you just use this method to do that? Agree we should have `playerList` correctly ordered so that the correct player should be first in the list, and the rest of the players are in the correct turn order, but this method could serve both purposes. I could see how it could be separate though, and not put the first move logic within here.

At the time this `Game` object is constructed, the `currentPlayersTurn` will be null I think, since we haven't set it yet. After that, yes, it'll be not null for the remainder of the game.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Turn.java:
```diff
@@ -0,0 +1,6 @@
+package com.debugmafia.clueless.state;
+
+class Turn
```
To answer one of the comments made about what parameters should the `Turn` class take - at a bare minimum it _has_  to take a `Player` instance (can't have a turn without a player). In the design document, we also have a `Set` of `BoardLocations` as a parameter...although I can't think at the moment as to why that is.

greenca6 replied:
The `Turn` class  will also just be a POJO. No other methods other than getters and setters.

greenca6 reviewed (commented):
See comments. On the right track for sure.

Sweethezus committed:
Addressed Review Comments

Addressed comments from PR review

Sweethezus committed:
Game update

Updated function calls and methods within Game to near completion

emtuls committed:
added setter method for setting available locations and finished endTurn method in Game.Java

Sweethezus commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -123,8 +123,7 @@ public Game endTurn(Player p) {
     Set<BoardLocation> openAdjacentLocations = this.board.getOpenAdjacentLocations(location);
-    // TODO: Set the available locations property on the current turn object to the
-    // result from above
+    currentPlayersTurn.setAvailableLocations(openAdjacentLocations);
```
These two lines can be combined into one.

Sweethezus reviewed (changes requested):
In general, we also need to have a way to "deactivate" a player. I think this should be done by having a get/set active in the Player class. This should be done before the PR is merged.

emtuls committed:
mostly finished Game.java, needs look over

Sweethezus committed:
Location check

Added check to make sure piece is in right location before suggestion is allowed

Sweethezus committed:
Merge branch 'develop' into feature/state

Sweethezus committed:
Updated Game logic

Updated the logic in Game so that we have a list of all players, primarily used for rebuttal purposes. When a player is removed, they are no longer in the activePlayers, but will still be in the allPlayers to be able to refute a claim.

Sweethezus committed:
Fix compilation

Fixed silly compilation mistake
